### PR TITLE
Refactor SASSD

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Modification

Refactor SASSD to be compatible with dev-1.x

The PAWarp is not supported yet in this implementation.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by ZCMax** [comment]
we need to add ut for sass

File: mmdet3d/models/detectors/sassd.py [hunk 1]
```diff
@@ -15,121 +15,68 @@ class SASSD(SingleStage3DDetector):
...
-        return voxels, num_points, coors_batch
-
-    def forward_train(self,
-                      points,
-                      img_metas,
-                      gt_bboxes_3d,
-                      gt_labels_3d,
-                      gt_bboxes_ignore=None):
-        """Training forward function.
+    def loss(self, batch_inputs_dict: dict, batch_data_samples: SampleList,
```
Comment (ZCMax) [inline mmdet3d/models/detectors/sassd.py:56]: Union[dict, list] -> dict ?

File: mmdet3d/models/detectors/sassd.py [hunk 2]
```diff
@@ -15,121 +15,68 @@ class SASSD(SingleStage3DDetector):
...
-            points (list[torch.Tensor]): Point cloud of each sample.
-            img_metas (list[dict]): Meta information of each sample
-            gt_bboxes_3d (list[:obj:`BaseInstance3DBoxes`]): Ground truth
-                boxes for each sample.
-            gt_labels_3d (list[torch.Tensor]): Ground truth labels for
-                boxes of each sampole
-            gt_bboxes_ignore (list[torch.Tensor], optional): Ground truth
-                boxes to be ignored. Defaults to None.
+            batch_inputs_dict (dict): The model input dict which include
+                'points', 'img' keys.
```
Comment (ZCMax) [inline mmdet3d/models/detectors/sassd.py:62]: `batch_inputs_dict` of `SASSD` should not include `img` keys?

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 3]
```diff
@@ -333,13 +341,16 @@ def forward(self, voxel_features, coors, batch_size, test_mode=False):
 
         return spatial_features, point_misc
 
-    def get_auxiliary_targets(self, nxyz, gt_boxes3d, enlarge=1.0):
+    def get_auxiliary_targets(self,
+                              nxyz: Tensor,
+                              gt_boxes3d: List[BaseInstance3DBoxes],
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:346]: rename `gt_boxes3d` to `gt_bboxes_3d` following other algorithms

Comment (ZCMax) [reply mmdet3d/models/middle_encoders/sparse_encoder.py:346]: please also check other places like `boxes_3d`

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 4]
```diff
@@ -366,7 +377,8 @@ def get_auxiliary_targets(self, nxyz, gt_boxes3d, enlarge=1.0):
 
         return pts_labels, center_offsets
 
-    def calculate_pts_offsets(self, points, boxes):
+    def calculate_pts_offsets(self, points: Tensor,
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:380]: boxes -> bboxes_3d

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 5]
```diff
@@ -404,7 +416,8 @@ def calculate_pts_offsets(self, points, boxes):
                         points[j][2] - (boxes[i][2] + boxes[i][2] / 2.0))
         return pts_indices.cpu(), center_offsets.cpu()
 
-    def aux_loss(self, points, point_cls, point_reg, gt_bboxes):
+    def aux_loss(self, points: Tensor, point_cls: Tensor, point_reg: Tensor,
+                 gt_bboxes: Tensor) -> dict:
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:420]: gt_bboxes_3d

File: mmdet3d/models/detectors/sassd.py [hunk 6]
```diff
@@ -15,121 +15,68 @@ class SASSD(SingleStage3DDetector):
...
-        voxels, num_points, coors = self.voxelize(points)
-        voxel_features = self.voxel_encoder(voxels, num_points, coors)
-        batch_size = coors[-1, 0].item() + 1
-        x, point_misc = self.middle_encoder(voxel_features, coors, batch_size,
+        voxel_dict = batch_inputs_dict['voxels']
+        voxel_features = self.voxel_encoder(voxel_dict['voxels'],
+                                            voxel_dict['num_points'],
+                                            voxel_dict['coors'])
+        batch_size = voxel_dict['coors'][-1, 0].item() + 1
+        x, point_misc = self.middle_encoder(voxel_features,
```
Comment (VVsssssk) [inline mmdet3d/models/detectors/sassd.py:48]: I suggest renaming `point_misc`.

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 7]
```diff
@@ -404,7 +416,8 @@ def calculate_pts_offsets(self, points, boxes):
                         points[j][2] - (boxes[i][2] + boxes[i][2] / 2.0))
         return pts_indices.cpu(), center_offsets.cpu()
 
-    def aux_loss(self, points, point_cls, point_reg, gt_bboxes):
+    def aux_loss(self, points: Tensor, point_cls: Tensor, point_reg: Tensor,
```
Comment (VVsssssk) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:419]: Whether get aux loss from a middle_enoder module normal? I think middle_encoder is a 3d backbone. So we need to put get aux loss logic to a head module?

Comment (JingweiZhang12) [reply mmdet3d/models/middle_encoders/sparse_encoder.py:420]: The input of aux loss is the temp variables in `middle_encoder` and they are not passed into the following pipeline like other head loss: `backbone` and `neck`. Also, as a characteristic of SASSD, it deeply binds with the middle_encoder to learn better point encoding.  Maybe it's unnecessary to add an aux_head directly upon the middle_encoder.

File: mmdet3d/models/detectors/sassd.py (hunk 6 above)
Comment (JingweiZhang12) [reply mmdet3d/models/detectors/sassd.py:48]: It's a little tricky because it's a tuple containing three different variables.  I add a comment to explain it.

File: mmdet3d/models/detectors/sassd.py [hunk 8]
```diff
@@ -1,135 +1,86 @@
...
+            init_cfg=init_cfg)
 
-        self.voxel_layer = Voxelization(**voxel_layer)
         self.voxel_encoder = MODELS.build(voxel_encoder)
         self.middle_encoder = MODELS.build(middle_encoder)
 
-    def extract_feat(self, points, img_metas=None, test_mode=False):
+    def extract_feat(self,
+                     batch_inputs_dict: dict,
+                     test_mode: bool = True) -> Tuple[Tensor]:
```
Comment (ZCMax) [inline mmdet3d/models/detectors/sassd.py:40]: add docstring here

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 9]
```diff
@@ -264,7 +268,11 @@ def __init__(self,
         self.point_cls = nn.Linear(64, 1, bias=False)
         self.point_reg = nn.Linear(64, 3, bias=False)
 
-    def forward(self, voxel_features, coors, batch_size, test_mode=False):
+    def forward(self,
+                voxel_features: Tensor,
+                coors: Tensor,
+                batch_size: Tensor,
+                test_mode: bool = False) -> Tuple[dict, tuple]:
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:275]: check `spatial_features` is dict or tensor

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 10]
```diff
@@ -333,13 +341,16 @@ def forward(self, voxel_features, coors, batch_size, test_mode=False):
...
+                              nxyz: Tensor,
+                              gt_boxes_3d: List[BaseInstance3DBoxes],
+                              enlarge: int = 1.0) -> Tuple[Tensor, Tensor]:
         """Get auxiliary target.
 
         Args:
             nxyz (torch.Tensor): Mean features of the points.
-            gt_boxes3d (torch.Tensor): Coordinates in shape (N, 4),
-                the columns in the order of (batch_idx, z_idx, y_idx, x_idx).
+            gt_boxes_3d (list[:obj:`BaseInstance3DBoxes`]):  Ground truth
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:352]: gt_bboxes_3d

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 11]
```diff
@@ -333,13 +341,16 @@ def forward(self, voxel_features, coors, batch_size, test_mode=False):
...
 
-    def get_auxiliary_targets(self, nxyz, gt_boxes3d, enlarge=1.0):
+    def get_auxiliary_targets(self,
+                              nxyz: Tensor,
+                              gt_boxes_3d: List[BaseInstance3DBoxes],
+                              enlarge: int = 1.0) -> Tuple[Tensor, Tensor]:
         """Get auxiliary target.
 
         Args:
             nxyz (torch.Tensor): Mean features of the points.
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:351]: rename `nxyz` to `points_feats` or some other clear name

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 12]
```diff
@@ -383,13 +395,13 @@ def calculate_pts_offsets(self, points, boxes):
                 if it belows to the box, with the shape of (M, 3).
                 Default background = 0.
         """
-        boxes_num = len(boxes)
+        boxes_num = len(bboxes_3d)
         pts_num = len(points)
         points = points.cuda()
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:400]: It's strange here, since points should be already put on CUDA?

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 13]
```diff
@@ -366,13 +377,14 @@ def get_auxiliary_targets(self, nxyz, gt_boxes3d, enlarge=1.0):
...
         return pts_labels, center_offsets
 
-    def calculate_pts_offsets(self, points, boxes):
+    def calculate_pts_offsets(self, points: Tensor,
+                              bboxes_3d: Tensor) -> Tuple[Tensor, Tensor]:
         """Find all boxes in which each point is, as well as the offsets from
         the box centers.
 
         Args:
             points (torch.Tensor): [M, 3], [x, y, z] in LiDAR/DEPTH coordinate
```
Comment (ZCMax) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:386]: [x, y, z] in LiDAR coordinate.

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 14]
```diff
@@ -383,13 +395,13 @@ def calculate_pts_offsets(self, points, boxes):
...
-        boxes_num = len(boxes)
+        boxes_num = len(bboxes_3d)
         pts_num = len(points)
         points = points.cuda()
-        boxes = boxes.to(points.device)
+        bboxes_3d = bboxes_3d.to(points.device)
 
-        box_idxs_of_pts = points_in_boxes_all(points[None, ...], boxes[None,
-                                                                       ...])
+        box_idxs_of_pts = points_in_boxes_all(points[None, ...],
```
[the remark this answers is not in the document]
Comment (ZCMax) [reply mmdet3d/models/middle_encoders/sparse_encoder.py:403]: please use `points_in_boxes_all` of `BaseInstance3DBoxes`

**Review by ZCMax — APPROVED** [review]

**Review by VVsssssk — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]
